### PR TITLE
Feat/task done snack undo

### LINKED ISSUE #7133: 💡 Undo mark as completed
### Problem Statement

If I accidentaly mark a task as complete, I cannot figure out which task I marked as completed to quickly undo the action.

### Possible Solution

I would propose a small notification in the bottom of the main app to permit to undo the action if it happened by accident. It can be a small timeout of 10 seconds to undo the action like in Gmail or other software.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### documentation gap

- [ ] ... related docs should be updated as part of the same PR whenever possible.
- [ ] ... I can help improve the documentation.

### PR DESCRIPTION
**Pull request by symonbaikov**
## Problem

  Closes #7133

  When a user accidentally marks a task as completed, there's no way to quickly undo the action — they have to manually find the task and uncheck it.

  ## Solution

  Added a snackbar notification that appears when a task is marked as done. It shows the task title and an **Undo** button that reverts the action within 10 seconds (similar to Gmail's send
  undo). The snackbar is styled to match the app's dark theme.

  ## Type of Change

  - [ ] Bug fix
  - [x] New feature
  - [ ] Breaking change
  - [ ] Refactoring
  - [ ] Documentation
  - [ ] Other (please describe)

  ## Checklist

  - [ ] I have included relevant changes to the documentation/[wiki](https://github.com/super-productivity/super-productivity/tree/master/docs/wiki).
  - [x] I have run `npm run checkFile` on changed `.ts`/`.scss` files
  - [ ] I have added tests for my changes (if applicable)
  - [x] Existing tests still pass
  - [x] My commit messages follow the Angular format (`type(scope): description`)
Screenshot From 2026-04-07 07-51-43

### CODE REVIEW

File: src/app/features/tasks/task/task.component.html [hunk 1]
```diff
@@ -223,6 +223,12 @@
         [progress]="progress()"
       ></progress-bar>
     }
+    @if (t.subTasks?.length) {
+      <progress-bar
+        [cssClass]="'bg-success sub-task-progress'"
+        [progress]="subTaskProgress()"
+      ></progress-bar>
+    }
```
[the remark this answers is not in the document]
Comment (symonbaikov) [reply src/app/features/tasks/task/task.component.html]: @copilot apply changes based on [this feedback](https://github.com/super-productivity/super-productivity/pull/7161#discussion_r3042986586)

File: src/assets/i18n/en.json [hunk 2]
```diff
@@ -1660,7 +1660,8 @@
         "REMINDER_DELETED": "Deleted reminder for task",
         "REMINDER_UPDATED": "Updated reminder for task \"{{title}}\"",
         "TASK_ALREADY_EXISTS": "Task <strong>{{title}}</strong> already exists",
-        "TASK_CREATED": "Created task \"{{taskTitle}}\""
+        "TASK_CREATED": "Created task \"{{taskTitle}}\"",
+        "TASK_DONE": "Marked \"{{title}}\" as done"
```
[the remark this answers is not in the document]
Comment (symonbaikov) [reply src/assets/i18n/en.json:1888]: @copilot apply changes based on [this feedback](https://github.com/super-productivity/super-productivity/pull/7161#discussion_r3042986575)

File: src/app/features/tasks/store/task-ui.effects.ts [hunk 3]
```diff
@@ -134,6 +135,35 @@ export class TaskUiEffects {
...
+                TaskSharedActions.updateTask({
+                  task: { id: task.id as string, changes: { isDone: false } },
+                }),
+              );
+            },
+          });
+        }),
+      ),
+    { dispatch: false },
+  );
```
[the remark this answers is not in the document]
Comment (symonbaikov) [reply src/app/features/tasks/store/task-ui.effects.ts:184]: @copilot apply changes based on [this feedback](https://github.com/super-productivity/super-productivity/pull/7161#discussion_r3042986564)

**Review by johannesjo — COMMENTED** [review]
## Review of PR #7161 — Feat/task done snack undo

Thanks for the contribution! The core "undo task done" snackbar feature is a nice UX improvement. I have some observations and issues to flag.

---

### Scope concern: unrelated sub-task progress bar changes

The PR title and description are about "task done snack undo," but roughly half the diff introduces an entirely separate feature — a **sub-task completion progress bar** on the task component (`task.component.html`, `task.component.ts`, `task.component.scss`). This includes:

- A new `subTaskProgress` computed signal
- A new `<progress-bar>` element in the template
- A new `hasVisibleTimeProgress` host class binding
- New SCSS for `.sub-task-progress`

These changes are unrelated to the undo snackbar and should be in a separate PR. Mixing unrelated features makes review harder and increases merge risk.

---

### Issues in `task-ui.effects.ts` — `snackDone$` effect

**1. Type guard narrows the wrong shape**

```ts
filter(
  ({ fullTask }): fullTask is Task => !!fullTask,
),
```

This is a type error. The type predicate `fullTask is Task` is applied to the destructured property, but `filter`'s type predicate must narrow the **entire value** flowing through the pipe, not a destructured field. The correct form would be:

```ts
filter(
  (val): val is { task: typeof val.task; fullTask: Task } => !!val.fullTask,
),
```

Or more simply, just use a regular filter and a non-null assertion / separate null check downstream.

**2. Snackbar fires for every `updateTask` where `isDone === true`, including sub-tasks**

When a parent task is marked done, all its sub-tasks are also marked done (via reducers or effects). This will trigger **multiple snackbar notifications** — one for the parent and one for each sub-task. Consider filtering to only show the snackbar for the task the user explicitly toggled (e.g., by checking if the task is a parent or by debouncing/deduplicating).

**3. The undo action dispatches `updateTask` with `isDone: false`, which will not fully reverse the original action**

Marking a task as done can trigger side effects beyond just setting `isDone` (e.g., updating sub-tasks, moving the task out of "current," playing a done sound). Simply dispatching `updateTask({ changes: { isDone: false } })` does not undo those cascading effects. Consider whether a more comprehensive undo mechanism is needed, or document this as a known limitation.

**4. No `selectTaskFeatureState` import added to the file**

Looking at the diff, `selectTaskFeatureState` is added to the imports from `./task.selectors`, but the original file (line 22-23) only imported `selectCurrentTask`, `selectCurrentTaskId`, and `selectUnplannedDeadlineTasksForToday`. The diff shows the import is added correctly — this is fine.

---

### Snackbar styling change is global

```scss
mat-snack-bar-container {
  --mat-snack-bar-container-color: var(--bg-lightest);
  --mat-snack-bar-supporting-text-color: var(--text-color);
}
```

This changes the styling of **all** snackbars in the application, not just the "task done" one. This may be intentional (and might even be an improvement), but it should be called out explicitly since it affects every existing snackbar. Has this been tested with success/error/warning snackbar types to ensure they still look correct?

---

### Translation string

The new translation `"TASK_DONE": "Marked \"{{title}}\" as done"` looks good and follows the existing pattern. Only `en.json` was edited, which is correct per project guidelines.

---

### Missing tests

The existing `task-ui.effects.spec.ts` has tests for other effects (`taskCreatedSnack$`, `snackDelete$`). The new `snackDone$` effect should have corresponding unit tests covering:
- Snackbar is shown when a task is marked done
- Snackbar is NOT shown for other `updateTask` changes (e.g., title change)
- The undo action dispatches the correct action
- (Ideally) behavior with sub-tasks

---

### Summary

| Aspect | Assessment |
|---|---|
| LOCAL_ACTIONS usage | Correct — uses `this._actions$` which is `inject(LOCAL_ACTIONS)` |
| State mutation | Correct — dispatches actions, no direct mutation |
| TypeScript typing | Type guard in filter is incorrect (see above) |
| Translation | Correctly added only to `en.json` |
| Scope | Sub-task progress bar changes should be a separate PR |
| Tests | Missing — should add unit tests for the new effect |
| Global styling | Snackbar style change affects all snackbars — needs verification |
| UX edge case | Multiple snackbars when parent + sub-tasks all marked done |

Recommendation: Please split out the sub-task progress bar into its own PR, fix the type guard, address the multiple-snackbar issue for sub-tasks, and add unit tests for the new effect.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by johannesjo** [comment]
Thanks — an undo on task completion is wanted (#7133), and the effect-based approach mirrors the existing `snackDelete$` nicely. Several blockers before merge:

1. **Build is broken.** `filter(({ fullTask }): fullTask is Task => !!fullTask)` in `task-ui.effects.ts` is a TS1230 error ("a type predicate cannot reference an element in a binding pattern") under strict mode. Use a non-destructured guard, e.g. `filter((v): v is { task: ...; fullTask: Task } => !!v.fullTask)`, or a plain null-check.

2. **Duplicate / wrong-target snack.** The effect keys off any `updateTask` with `isDone: true`, so auto-mark-parent-as-done (`task-internal.effects.ts`) fires a second snackbar whose Undo reverts the parent, not the subtask the user clicked. Please gate on genuine user intent (the explicitly toggled task, or a skip-snack meta flag on the internal dispatch).

3. **Scope creep.** Roughly half the diff adds an unrelated subtask progress bar to the `task.component.*` hot path — please split that into its own PR.

4. **Global snackbar restyle.** The `mat-snack-bar-container` override in `_overwrite-material.scss` affects every snackbar app-wide (success/error/warning) — scope it or confirm it's intended.

5. **No tests**, and the branch currently conflicts with master.

The core idea is good; mostly needs a split + the compile fix + a user-intent gate.

**Comment by SilverGreen93** [comment]
Tested on the web interface, works exactly as requested! Thanks!

**Comment by johannesjo** [comment]
Hey @symonbaikov — first off, thank you for the work here, and sorry for the confusion and the back-and-forth on this one. 🙏 You clearly put real care into it (the tests and the skip-flag for the subtask→parent cascade especially), and that's appreciated.

After thinking it through, I've decided not to merge this. The main reason: in the default work view, done tasks stay visible in the list with their checkbox right there, so the "undo" already exists one click away. An always-on snackbar that fires on *every* completion (and on all the other paths that set a task done — keyboard, daily summary, finish-day, repeat tasks, etc.) ends up being a fairly intrusive interruption for something that's already trivially reversible, especially for power users who check off lots of tasks a day.

So this is much more a product-direction call than anything about the quality of your PR. Closing it for that reason — but genuinely thank you for contributing, and I hope you'll send more. 🙇

**Comment by SilverGreen93** [comment]
> After thinking it through, I've decided not to merge this. The main reason: in the default work view, done tasks stay visible in the list with their checkbox right there, so the "undo" already exists one click away [...]  already trivially reversible

There is nothing trivial about reversing the completion checkbox if you have no idea on which task you accidentally marked as completed. If you can suggest a better alternative, I am open for discussion.

On Android, I mistakenly click on tasks by accident or just swipe by accident marking them completed when I don't want to, so the Undo feature would be really the one-click-away solution.

**Comment by symonbaikov** [comment]
@johannesjo Of course! I'm just clearing out here old pull requests to focus on new ones. Soon will be more!
